src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -1079,10 +1086,13 @@ trait Contexts { self: Analyzer =>
         defSym = lookupInScope(cx.scope) match {
           case Nil                  => searchPrefix
           case entries @ (hd :: tl) =>
-            // we have a winner: record the symbol depth
-            symbolDepth = (cx.depth - cx.scope.nestingLevel) + hd.depth
-            if (tl.isEmpty) hd.sym
-            else newOverloaded(cx.owner, pre, entries)
+            if (hd.owner eq enclosingPatternCx.scope) NoSymbol
```

Comment:
Agreed, this is pretty messy.

Typechecking patterns can create contexts that are nested in the `CaseDef` context. If we start outside the pattern context, we miss the existentials. This PR opts to leave the context chain alone, but skip symbols directly defined in the pattern context.

It might, however, be cleaner to typecheck the pattern with a new context whose outer context skips the `CaseDef` context.

So instead of:

```scala
      val pat1 = {
        context.pattern = true
        try typedPattern(cdef.pat, pattpe.withoutAnnotations)
        finally context.pattern = false
      }
```

```scala
      val pat1 = {
        val patternTyper = newTyper(context.outer.make(....))
        patternTyper.typedPattern(cdef.pat, pattpe.withoutAnnotations)
      }
```

We just need to cautious to ensure that while typechecking the pattern we never need to do something with `context.enclosingCaseDef` (e.g. work with GADT bounds in `savedTypeBounds`).

I'll come try a few alternatives and report back, but that might be after Scaladays. I'll close this for now.